### PR TITLE
Enhancement: 1. `go mod tidy` + 2. `diffutils` when mac OS

### PR DESCRIPTION
**Pull request by tzaffi**
## Summary

Introducing `make tidy` and enforcing it in `codegen_verification.sh` as well as `make sanity`.

As `go mod tidy`'s behavior can change from version to version of `go`, `make tidy` insists that the expected go build major version be applied.

(The build version is defined in `scripts/get_golang_version.sh`).

Also adding `diffutils` to `configure_dev.sh` mac `install_or_upgrade()`'s (see the [comment below](https://github.com/algorand/go-algorand/pull/5398#discussion_r1204832324) for an explanation)

## Test Plan

CI and locally to ensure that `make tidy` works when the expected go-version hasn't already been installed.

### CODE REVIEW

**Comment by tzaffi** [comment]
> You could also have this implemented with a few lines of circleci YAML or github actions YAML — I'm not sure that writing a custom Go program has any significant advantage over the shell and some CI glue e.g. https://github.com/katexochen/go-tidy-check/blob/main/action.yml

Purely experimental at this point. The one advantage of using `go` is access to the `golang.org/x/mod/modfile.File` struct which is a bit less hackey than `grep`'ing in a shell script in case you want to iterate over the current packages.

File: Makefile [hunk 1]
```diff
@@ -101,6 +105,12 @@ fix: build
 lint: deps
 	$(GOPATH1)/bin/golangci-lint run -c .golangci.yml
 
+tidy:
+	export PATH=$(GOPATH1)/bin:$(PATH) && \
+	go install golang.org/dl/go$(GOLANG_VERSION_BUILD)@latest && \
```
[the remark this answers is not in the document]
Comment (tzaffi) [reply Makefile]: see e655e814960fd7e7990311674f87645aca395cb6

File: Makefile [hunk 2]
```diff
@@ -101,6 +107,18 @@ fix: build
 lint: deps
 	$(GOPATH1)/bin/golangci-lint run -c .golangci.yml
 
+# NOTE: tidy will download and install go version GOLANG_VERSION_BUILD 
+# 		if this isn't the system's current version
```
Comment (winder) [inline Makefile:111]: It's nice that `go install` can manage this, but it still seems a little strange to me that this is done in the Makefile.

I'm not completely oppose to the change here, but exiting with an error that tells you the go version is wrong would be my preference. Maybe even split it out with a `check` target which checks your environment, and add that as a dependency to other things which might be wrongish with the wrong version of go installed.

File: Makefile (hunk 1 above)
[the remark this answers is not in the document]
Comment (gmalouf) [reply Makefile]: I strongly prefer not downloading - I manage golang via my system build env.

File: scripts/travis/codegen_verification.sh [hunk 3]
```diff
@@ -61,5 +61,17 @@ else
...
+make tidy
+if [[ -n $(git status --porcelain) ]]; then
+   echo Dirty after go mod tidy - did you forget to run make tidy?
+   git status -s
+   git --no-pager diff
+   exit 1
+else
+   echo All tidy
+fi
+
```
Comment (tzaffi) [inline scripts/travis/codegen_verification.sh:74]:
```suggestion
```

File: Makefile (hunk 1 above)
[the remark this answers is not in the document]
Comment (tzaffi) [reply Makefile]: see c7e31fe657df6d31e708b228b4f4f6ac6246b457

File: Makefile (hunk 2 above)
Comment (tzaffi) [reply Makefile]: c7e31fe657df6d31e708b228b4f4f6ac6246b457

File: Makefile [hunk 4]
```diff
@@ -101,6 +107,18 @@ fix: build
...
+tidy:
+	@if [ $(CURRENT_GO_VERSION) != $(GOLANG_VERSION_BUILD) ]; then \
+		export PATH=$(GOPATH1)/bin:$(PATH) && \
+		go install golang.org/dl/go$(GOLANG_VERSION_BUILD)@latest && \
+		go$(GOLANG_VERSION_BUILD) download && \
+		go$(GOLANG_VERSION_BUILD) mod tidy -compat=$(GOLANG_VERSION_SUPPORT); \
+	else \
+		go mod tidy -compat=$(GOLANG_VERSION_SUPPORT); \
+	fi
```
[the remark this answers is not in the document]
Comment (tzaffi) [reply Makefile]: c7e31fe657df6d31e708b228b4f4f6ac6246b457

**Review by winder — APPROVED** [review]

**Review by gmalouf — CHANGES_REQUESTED** [review]
I'm mostly good with this, think you should add a call to `tidy` under the `sanity` make alias (I think just before the `fix` call).

File: scripts/configure_dev.sh [hunk 5]
```diff
@@ -86,6 +86,7 @@ elif [ "${OS}" = "darwin" ]; then
         install_or_upgrade automake
         install_or_upgrade python3
         install_or_upgrade lnav
+        install_or_upgrade diffutils
```
Comment (tzaffi) [inline scripts/configure_dev.sh:89]: I wasn't able to get `make lint` to pass locally until I installed `diffutils` per these links:
* [lakeFS forum discussion](https://forum.lakefs.io/t/10003060/beginners-question-the-linting-failed-on-https-github-com-tr)
* https://github.com/golangci/golangci-lint/issues/3087

File: Makefile [hunk 6]
```diff
@@ -13,6 +13,18 @@ ARCH        := $(shell ./scripts/archtype.sh)
...
+GOLANG_VERSION_SUPPORT		:= $(lastword $(GOLANG_VERSIONS))
+GOLANG_VERSION_BUILD_MAJOR	:= $(shell echo $(GOLANG_VERSION_BUILD) | cut -d'.' -f1,2)
+CURRENT_GO_VERSION			:= $(shell go version | cut -d " " -f 3 | tr -d 'go')
+CURRENT_GO_VERSION_MAJOR	:= $(shell echo $(CURRENT_GO_VERSION) | cut -d'.' -f1,2)
+
+GOLANGCILINT_VERSION := $(shell yq e '.jobs.reviewdog-errors.steps[] | select(.name == "reviewdog-golangci-lint").with.golangci_lint_version' .github/workflows/reviewdog.yml)
+
+echo:
+	@echo $(GOLANGCILINT_VERSION)
+
```
Comment (tzaffi) [inline Makefile:27]:
```suggestion
```

**Comment by tzaffi** [comment]
> I'm mostly good with this, think you should add a call to `tidy` under the `sanity` make alias (I think just before the `fix` call).

See the latest

**Review by gmalouf — APPROVED** [review]

**Review by winder — APPROVED** [review]

**Pull request merged**